presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -284,61 +285,68 @@ protected Type visitQualifiedNameReference(QualifiedNameReference node, Analysis
-
+            columnReferences.add(node);
             return field.getType();
         }
 
-        private Type tryVisitRowFieldAccessor(QualifiedNameReference node)
+        @Override
+        protected Type visitDereferenceExpression(DereferenceExpression node, AnalysisContext context)
         {
-            if (node.getName().getParts().size() < 2) {
-                throw createMissingAttributeException(node);
-            }
-            QualifiedName base = new QualifiedName(node.getName().getParts().subList(0, node.getName().getParts().size() - 1));
-            List<Field> matches = tupleDescriptor.resolveFields(base);
-            if (matches.isEmpty()) {
-                throw createMissingAttributeException(node);
-            }
-            if (matches.size() > 1) {
-                throw new SemanticException(AMBIGUOUS_ATTRIBUTE, node, "Column '%s' is ambiguous", node.getName());
-            }
+            QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
 
-            Field field = Iterables.getOnlyElement(matches);
-            if (field.getType() instanceof RowType) {
-                RowType rowType = checkType(field.getType(), RowType.class, "field.getType()");
-                Type rowFieldType = null;
-                for (RowField rowField : rowType.getFields()) {
-                    if (rowField.getName().equals(Optional.of(node.getName().getSuffix()))) {
-                        rowFieldType = rowField.getType();
-                        break;
-                    }
+            // If this Dereference looks like column reference, try match it to column first.
+            if (qualifiedName != null) {
+                List<Field> matches = tupleDescriptor.resolveFields(qualifiedName);
+                if (matches.size() > 1) {
+                    throw new SemanticException(AMBIGUOUS_ATTRIBUTE, node, "Column '%s' is ambiguous", node);
                 }
-                if (rowFieldType == null) {
-                    throw createMissingAttributeException(node);
+
+                if (matches.size() == 1) {
+                    Field field = Iterables.getOnlyElement(matches);
+                    int fieldIndex = tupleDescriptor.indexOf(field);
+                    resolvedNames.put(node, fieldIndex);
+                    expressionTypes.put(node, field.getType());
+                    columnReferences.add(node);
+                    return field.getType();
                 }
-                int fieldIndex = tupleDescriptor.indexOf(field);
-                resolvedNames.put(node.getName(), fieldIndex);
-                expressionTypes.put(node, rowFieldType);
-                rowFieldReferences.put(node, true);
+            }
+
+            Type baseType = process(node.getBase(), context);
+            if (!(baseType instanceof RowType)) {
+                throw new SemanticException(SemanticErrorCode.TYPE_MISMATCH, node.getBase(), "Expression %s is not RowType", node.getBase());
```

Comment:
I think the user will see this message, so change it to say "Expression %s is not of type ROW".